accumulator_updater/programs/accumulator_updater/src/state/accumulator_input.rs

```diff
@@ -43,48 +111,138 @@ impl AccumulatorInput {
         let expected_key = self.derive_pda(cpi_caller, base_account)?;
         require_keys_eq!(expected_key, key);
-        require_eq!(self.header.account_schema, account_schema);
         Ok(())
     }
+}
 
-    pub fn persist(&self, ai: &AccountInfo) -> Result<()> {
-        AccountSerialize::try_serialize(self, &mut &mut ai.data.borrow_mut()[..]).map_err(|e| {
-            msg!("original error: {:?}", e);
-            AccumulatorUpdaterError::SerializeError
-        })?;
-        Ok(())
+
+#[cfg(test)]
+mod test {
+    use {
+        super::*,
+        bytemuck::bytes_of,
+        std::mem::{
+            align_of,
+            size_of,
+        },
+    };
+
+    fn data_bytes(data: Vec<u8>) -> Vec<u8> {
+        let mut bytes = vec![];
+        for d in data {
+            bytes.extend_from_slice(&d.to_be_bytes());
+        }
+        bytes
     }
-}
 
-//TODO:
-// - implement custom serialization & set alignment
-// - what other fields are needed?
-#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug, Default)]
-pub struct AccumulatorHeader {
-    pub bump:           u8,
-    pub version:        u8,
-    pub account_schema: u8,
-}
 
+    #[test]
+    fn test_sizes_and_alignments() {
+        let (header_idx_size, header_idx_align) = (
+            size_of::<AccumulatorHeader>(),
+            align_of::<AccumulatorHeader>(),
+        );
 
-impl AccumulatorHeader {
-    pub const SIZE: usize = 1 + 1 + 1;
-    pub const CURRENT_VERSION: u8 = 1;
+        let (input_size, input_align) = (
+            size_of::<AccumulatorInput>(),
+            align_of::<AccumulatorInput>(),
+        );
 
-    pub fn new(bump: u8, account_schema: u8) -> Self {
-        Self {
-            bump,
-            version: Self::CURRENT_VERSION,
-            account_schema,
+
+        println!(
+            r"
+
+            header
+                size: {header_idx_size:?}
+                align: {header_idx_align:?}
+            input
+                size: {input_size:?}
+                align: {input_align:?}
+        "
+        );
+        assert_eq!(header_idx_size, 514);
+        assert_eq!(header_idx_align, 2);
+        assert_eq!(input_size, 10_232);
+        assert_eq!(input_align, 2);
+    }
+
+    #[test]
+    fn test_put_all() {
+        let data = vec![vec![12, 34], vec![56, 78, 90]];
+        let data_bytes: Vec<Vec<u8>> = data.into_iter().map(data_bytes).collect();
+
+        let accumulator_input = &mut AccumulatorInput::new(0);
+
+        let (num_msgs, num_bytes) = accumulator_input.put_all(&data_bytes);
+        assert_eq!(num_msgs, 2);
+        assert_eq!(num_bytes, 5);
+
+
+        assert_eq!(accumulator_input.header.end_offsets[0], 2);
+        assert_eq!(accumulator_input.header.end_offsets[1], 5);
+
+
+        let accumulator_input_bytes = bytes_of(accumulator_input);
+
+        // *note* minus 8 here since no account discriminator when using
+        // `bytes_of`directly on accumulator_input
```

Comment:
general note on wording your documentation: you tend to write comments in a way that assumes the reader has lots of context for the code. It's usually more helpful to explain things in a way that assumes the reader *doesn't* have context. For example, this could say: "The header_len field represents the size of all data prior to the message bytes. This includes the account discriminator, which is not part of the header struct. Subtract the size of the discriminator (8 bytes) to compensate."

(this is like exactly the kind of place where you want to put a comment, so my comment is about the wording.)